src/d3d12/d3d12_descriptors_allocations.cpp

```diff
@@ -0,0 +1,391 @@
+		{
+			// The previous block is exactly behind the block that is to be freed.
+			//
+			// PrevBlock.Offset           Offset
+			// |                          |
+			// |<-----PrevBlock.Size----->|<------Size-------->|
+			//
+
+			// Increase the block size by the size of merging with the previous block.
+			offset = prevBlockIt->first;
+			num_descriptors += prevBlockIt->second.m_size;
+
+			// Remove the previous block from the free list.
+			m_freelist_by_size.erase(prevBlockIt->second.m_freelist_by_size_itr);
+			m_freelist_by_offset.erase(prevBlockIt);
+		}
+
+		if (nextBlockIt != m_freelist_by_offset.end() &&
+			offset + num_descriptors == nextBlockIt->first)
+		{
+			// The next block is exactly in front of the block that is to be freed.
+			//
+			// Offset               NextBlock.Offset 
+			// |                    |
+			// |<------Size-------->|<-----NextBlock.Size----->|
+
+			// Increase the block size by the size of merging with the next block.
+			num_descriptors += nextBlockIt->second.m_size;
+
+			// Remove the next block from the free list.
+			m_freelist_by_size.erase(nextBlockIt->second.m_freelist_by_size_itr);
+			m_freelist_by_offset.erase(nextBlockIt);
+		}
+
+		// Add the freed block to the free list.
+		AddNewBlock(offset, num_descriptors);
+	}
+
+	void DescriptorAllocatorPage::ReleaseStaleDescriptors()
+	{
+		std::lock_guard<std::mutex> lock(m_allocation_mutex);
+
+		while (!m_stale_descriptors.empty() && m_stale_descriptors.front().m_frame_number <= m_render_system.GetFrameIdx())
+		{
+			auto& staleDescriptor = m_stale_descriptors.front();
+
+			// The offset of the descriptor in the heap.
+			auto offset = staleDescriptor.m_offset;
+			// The number of descriptors that were allocated.
+			auto numDescriptors = staleDescriptor.m_size;
+
+			FreeBlock(offset, numDescriptors);
+
+			m_stale_descriptors.pop();
+		}
+	}
+}
+
+//////////////////////////
+// DESCRIPTOR ALLOCATOR //
+//////////////////////////
+
+namespace wr
+{
+	DescriptorAllocator::DescriptorAllocator(D3D12RenderSystem& render_system, DescriptorHeapType type, uint32_t num_descriptors_per_heap)
+		: m_render_system(render_system)
+		, m_heap_type(type)
+		, m_num_descriptors_per_heap(num_descriptors_per_heap)
+	{
+	}
+
+	DescriptorAllocator::~DescriptorAllocator()
+	{}
+
+	std::shared_ptr<DescriptorAllocatorPage> DescriptorAllocator::CreateAllocatorPage()
+	{
+		auto new_page = std::make_shared<DescriptorAllocatorPage>(m_render_system, m_heap_type, m_num_descriptors_per_heap);
+
+		m_heap_pool.emplace_back(new_page);
+		m_available_heaps.insert(m_heap_pool.size() - 1);
+
+		return new_page;
+	}
+
+	DescriptorAllocation DescriptorAllocator::Allocate(uint32_t num_dscriptors)
+	{
+		std::lock_guard<std::mutex> lock(m_allocation_mutex);
+
+		DescriptorAllocation allocation;
+
+		for (auto iter = m_available_heaps.begin(); iter != m_available_heaps.end(); ++iter)
+		{
+			auto allocatorPage = m_heap_pool[*iter];
```

Comment:
When copy pasting code please change it to fit the style guide.

Reply:
👍